File: neoforge/src/main/java/com/github/mystery2099/voxlib/neoforge/VoxLibNeoForgeClient.java
package com.github.mystery2099.voxlib.neoforge;

import com.github.mystery2099.voxlib.config.VoxLibConfig;
import com.github.mystery2099.voxlib.debug.VoxelShapeDebugClient;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.screens.Screen;
import net.neoforged.fml.ModContainer;
import net.neoforged.fml.loading.FMLPaths;
import net.neoforged.neoforge.client.event.RenderHighlightEvent;
import net.neoforged.neoforge.client.gui.IConfigScreenFactory;
import net.neoforged.neoforge.common.NeoForge;

final class VoxLibNeoForgeClient {
    static void initialize(ModContainer modContainer) {
        VoxLibConfig.Companion.initialize(FMLPaths.CONFIGDIR.get());
        VoxelShapeDebugClient.INSTANCE.initialize();
        modContainer.registerExtensionPoint(IConfigScreenFactory.class,
            new IConfigScreenFactory() {
                @Override
                public Screen createScreen(Minecraft minecraft, Screen parent) {
                    return VoxLibConfig.Companion.createConfigScreen(parent);
                }
            });
        NeoForge.EVENT_BUS.addListener(VoxLibNeoForgeClient::renderBlockHighlight);
    }

    private static void renderBlockHighlight(RenderHighlightEvent.Block event) {
        var world = Minecraft.getInstance().level;
        if (world == null) return;
        var pos = event.getTarget().getBlockPos();
        var camera = event.getCamera();
        var cameraPos = camera.getPosition();
        boolean renderVanilla = VoxelShapeDebugClient.INSTANCE.renderTargetedShapes(
            event.getPoseStack(), event.getMultiBufferSource(), world, pos, world.getBlockState(pos),
            camera.getEntity(), cameraPos.x, cameraPos.y, cameraPos.z);
        if (!renderVanilla) event.setCanceled(true);
    }
}

File: neoforge/src/main/kotlin/com/github/mystery2099/voxlib/neoforge/VoxLibNeoForgeClient.kt
package com.github.mystery2099.voxlib.neoforge

import com.github.mystery2099.voxlib.config.VoxLibConfig
import com.github.mystery2099.voxlib.debug.VoxelShapeDebugClient
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.screens.Screen
import net.neoforged.fml.ModContainer
import net.neoforged.fml.loading.FMLPaths
import net.neoforged.neoforge.client.event.RenderHighlightEvent
import net.neoforged.neoforge.client.gui.IConfigScreenFactory
import net.neoforged.neoforge.common.NeoForge

internal object VoxLibNeoForgeClient {
    fun initialize(modContainer: ModContainer) {
        VoxLibConfig.initialize(FMLPaths.CONFIGDIR.get())
        VoxelShapeDebugClient.initialize()
        modContainer.registerExtensionPoint(
            IConfigScreenFactory::class.java,
            IConfigScreenFactory { _: Minecraft, parent: Screen ->
                VoxLibConfig.createConfigScreen(parent)
            }
        )
        NeoForge.EVENT_BUS.addListener { event: RenderHighlightEvent.Block ->
            renderBlockHighlight(event)
        }
    }

    private fun renderBlockHighlight(event: RenderHighlightEvent.Block) {
        val world = Minecraft.getInstance().level ?: return
        val pos = event.target.blockPos
        val camera = event.camera
        val cameraPos = camera.position
        val renderVanilla = VoxelShapeDebugClient.renderTargetedShapes(
            event.poseStack, event.multiBufferSource, world, pos, world.getBlockState(pos),
            camera.entity, cameraPos.x, cameraPos.y, cameraPos.z
        )
        if (!renderVanilla) event.isCanceled = true
    }
}
